include/dali/core/mm/cuda_vm_resource.h

```diff
@@ -0,0 +1,613 @@
+      assert(mapping.size() == mapped.size());
+      assert(mapping.size() == available.size());
+    }
+
+    void resize(int new_num_blocks) {
+      if (new_num_blocks < num_blocks()) {
+        int no_longer_in_range = 0;
+        for (int i = new_num_blocks; i < num_blocks(); i++) {
+          if (mapping[i]) {
+            cuvm::Unmap(block_ptr(i), block_size);
+            CUDA_DTOR_CALL(cuMemRelease(mapping[i]));
+            mapping[i] = {};
+            if (available[i])
+              no_longer_in_range++;
+          }
+        }
+        available_blocks -= no_longer_in_range;
+      }
+      mapping.resize(new_num_blocks);
+      mapped.resize(new_num_blocks);
+      available.resize(new_num_blocks);
+      address_range.size() = new_num_blocks * block_size;
+    }
+
+    void set_block_availability(char *start, char *end, bool availability) {
+      int blk_start = (start - block_ptr<char>(0)) / block_size;
+      int blk_end   = (end   - block_ptr<char>(0)) / block_size;
+      int flipped = 0;
+      for (int b = available.find(!availability, blk_start); b < blk_end; ) {
+        int first_flipped = available.find(availability, b+1);
+        flipped += first_flipped - b;
+        b = first_flipped + 1;
+      }
+      if (availability)
+        available_blocks += flipped;
+      else
+        available_blocks -= flipped;
+      available.fill(blk_start, blk_end, availability);
+    }
+
+    cuvm::CUAddressRange    address_range;
+    vector<mem_handle_t>    mapping;
+    bitmask                 mapped, available;
+    ptrdiff_t               available_blocks = 0;
+    size_t                  block_size = 0;
+  };
+
+  SmallVector<va_region, 8> va_regions_;
+  SmallVector<cuvm::CUMemAddressRange, 8> va_ranges_;
+  size_t initial_va_size_ = 0;
+  size_t block_size_ = 0;
+  size_t total_mem_ = 0;
+  int device_ordinal_ = -1;
+
+  void adjust_params(size_t &size, size_t &alignment) {
+    alignment = std::max(alignment, next_pow2(size >> 11));
+    alignment = std::min(alignment, block_size_);
+    size = align_up(size, alignment);
+  }
+
+  void *do_allocate(size_t size, size_t alignment) override {
+    adjust_params(size, alignment);
+    if (size == 0)
+      return nullptr;
+    lock_guard pool_guard(pool_lock_);
+    // try to get a free region that's already mapped
+    void *ptr = try_get_mapped(size, alignment);
+    if (ptr) {
+      stat_add_allocation(size);
+      return ptr;
+    }
+    DeviceGuard dg(device_ordinal_);
+    mem_lock_guard mem_guard(mem_lock_);
+    ptr = get_va(size, alignment);
+    map_storage(ptr, size);
+    ptr = try_get_mapped(size, alignment);
+    assert(ptr != nullptr);
+    stat_add_allocation(size);
+    return ptr;
+  }
+
+  void *try_get_mapped(size_t size, size_t alignment) {
+    char *ptr = static_cast<char*>(free_mapped_.get(size, alignment));
+    if (ptr) {
+      stat_take_free(size);
+      free_va_.get_specific_block(ptr, ptr + size);
+      mark_as_unavailable(ptr, size);
```

Comment:
I beg to differ. I want reporting the change to stat to be the first thing after that change happened. In this case, the change to the amount of free memory happens outside the `if` - in `free_mapped_.get(size, alignment)`. When that line succeeds, I report the change to stat, not risking that something in the meantime will throw and would destroy the consistency of `stat` - and at least theoretically, `free_va_.get_specific_block()` can throw because it can split a free block, which in turn might result in heap allocation - which can end in `bad_alloc` begin thrown.